browser/greaselion/greaselion_tab_helper.cc

```diff
@@ -0,0 +1,89 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/greaselion/greaselion_tab_helper.h"
+
+#include <string>
+#include <vector>
+
+#include "base/bind_helpers.h"
+#include "base/strings/utf_string_conversions.h"
+#include "brave/browser/greaselion/greaselion_service_factory.h"
+#include "brave/common/brave_isolated_worlds.h"
+#include "brave/components/brave_rewards/browser/rewards_service.h"
+#include "brave/components/brave_rewards/browser/rewards_service_factory.h"
+#include "brave/components/greaselion/browser/greaselion_service.h"
+#include "chrome/browser/profiles/profile.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_user_data.h"
+
+using greaselion::GreaselionService;
+using greaselion::GreaselionServiceFactory;
+
+namespace greaselion {
+
+GreaselionTabHelper::GreaselionTabHelper(content::WebContents* web_contents)
+    : WebContentsObserver(web_contents) {
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents->GetBrowserContext());
+  if (profile) {
+    brave_rewards::RewardsService* rewards_service =
+        brave_rewards::RewardsServiceFactory::GetForProfile(profile);
+    if (rewards_service)
+      rewards_service->AddObserver(this);
+  }
+}
+
+GreaselionTabHelper::~GreaselionTabHelper() {
+  if (!web_contents())
```

Comment:
I'm a bit worried about this because it's possible to call `tab_helper->Observe(nullptr)` which would then prevent the observer from being removed and potentially cause a crash. If the rewards service exists it is guaranteed to outlive any open tabs so it's safe to cache `rewards_service_` and just check that for null here